R/fair-aaa.R

```diff
@@ -0,0 +1,137 @@
+#'
+#' @section Relevant Group Level:
+#' By default,
+#'
+#' Additional arguments can be passed to the function outputted by
+#' the function that this function outputs. That is:
+#'
+#' ```
+#' res_fairness <- fairness_metric(...)
+#' res_by <- res_fairness(by)
+#' res_by(..., additional_arguments_to_.post = TRUE)
+#' ```
+#'
+#' For finer control of how groups in `by` are treated, use the
+#' `.post` argument.
+#'
+#' @return A function with one argument, `by`, indicating the data-masked
+#' variable giving the sensitive feature. See the documentation on the
+#' return value of fairness metrics like [demographic_parity()],
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
+#'
+#' # `demographic_parity`, among other fairness metrics,
+#' # is generated with `fairness_metric()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#' demographic_parity_ <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity",
+#'     .post = diff_range
+#'   )
+#'
+#' m_set <- metric_set(demographic_parity_(Resample))
+#'
+#' m_set(hpc_cv, truth = obs, estimate = pred)
+#'
+#' # the `post` argument can be used to accommodate a wide
+#' # variety of parameterizations. to encode demographic
+#' # parity as a ratio inside of a difference, for example:
+#' ratio_range <- function(x, ...) {
+#'   range <- range(x$.estimate)
+#'   range[1] / range[2]
+#' }
+#'
+#' demographic_parity_ratio <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity_ratio",
+#'     .post = ratio_range
+#'   )
+#'
+#' @export
+fairness_metric <- function(.fn, .name, .post) {
+  if (rlang::is_missing(.fn) || !inherits_any(.fn, c("metric", "metric_set"))) {
+    rlang::abort("`.fn` must be a metric function or metric set.")
+  }
+  if (rlang::is_missing(.name) || !is_string(.name)) {
+    abort("`.name` must be a string.")
+  }
+  if (rlang::is_missing(.post) || !is_function(.post)) {
+    abort("`.post` must be a function.")
+  }
+
+  function(by) {
+    by_str <- rlang::as_string(rlang::enexpr(by))
+    res <-
+      function(data, ...) {
+        gp_vars <- dplyr::group_vars(data)
+
+        res <- dplyr::group_by(data, {{by}}, .add = TRUE)
+        res <- .fn(res, ...)
+
+        if (length(gp_vars) > 0) {
+          splits <- vctrs::vec_split(res, res[gp_vars])
+          .estimate <- vapply(splits$val, .post, numeric(1), ...)
+        } else {
+          .estimate <- .post(res, ...)
+        }
+
+        if (!rlang::is_bare_numeric(.estimate)) {
+          abort(
+            "`.post` must return a single numeric value.",
+            call = rlang::call2("fairness_metric")
+          )
+        }
+
+        if (length(gp_vars) > 0) {
+          res <- dplyr::group_by(res, !!!dplyr::groups(data), .add = FALSE)
+        }
+
+        res <-
+          dplyr::summarize(
```

Comment:
This might take a little time to convert, but I think we should avoid calling `dplyr::summarize()` directly, and instead use loop with the dplyr group helpers. Similar to how it was done in https://github.com/tidymodels/yardstick/pull/405. I'm imagining that it would make the errors better

Reply:
I'm putting it here, since this issue should be resolved first. The current implementation doesn't give good errors 😞  I'm very positive we can make it better!

``` r
library(yardstick)

data(hpc_cv)

equalized_odds_resample <- equalized_odds(Resample)

hpc_cv |>
  equalized_odds_resample(truth = VF, estimate = pred)
#> Error in `metric_set()`:
#> ! Failed to compute `sens()`.
#> Caused by error:
#> ! `truth` should be a factor, not a `numeric`.
```

<sup>Created on 2023-05-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Transitioned from `summarize()` in [27f53e9](https://github.com/tidymodels/yardstick/pull/434/commits/27f53e9fed2ac1627702b267bd6d6177cfdb5653)! Thought that error message still looks the same.

Reply:
Errors from `demographic_parity()` and `equal_opportunity()` look a bit different since their `.fn` is a metric rather than a metric set. With those:

``` r
library(yardstick)

data(hpc_cv)

equal_opportunity_resample <- equal_opportunity(Resample)

hpc_cv |>
  equal_opportunity_resample(truth = VF, estimate = pred)
```

It'd be nice if that could say `equal_opportunity_resample()` instead of `.fn()`, since the user maybe need not know about the existence of `fairness_metric()`, but this might require some cleverness in passing the `call` to metric functions like `sens()` or `spec()`.

Reply:
I fully support trying to apply cleverness! 😄

Reply:
Addressed in [375269b](https://github.com/tidymodels/yardstick/pull/434/commits/375269bdf88f6e494f4506cb282d2e292dfef40f)! Thanks for the nudge on this. :)

``` r
library(yardstick)

data(hpc_cv)

equalized_odds_resample <- equalized_odds(Resample)

hpc_cv |>
  equalized_odds_resample(truth = VF, estimate = pred)
#> Error in `equalized_odds()`:
#> ! `truth` should be a factor, not a `numeric`.
```

<sup>Created on 2023-06-21 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>